FirebaseSwiftUI/FirebaseAuthSwiftUI/Sources/Services/AuthService.swift

```diff
@@ -0,0 +1,343 @@
+      }
+      return provider
+    }
+  }
+
+  private var safePhoneAuthProvider: PhoneAuthProviderProtocol {
+    get throws {
+      guard let provider = phoneAuthProvider else {
+        throw AuthServiceError
+          .notConfiguredProvider("`PhoneAuthProviderSwift` has not been configured")
+      }
+      return provider
+    }
+  }
+
+  private func safeActionCodeSettings() throws -> ActionCodeSettings {
+    // email sign-in requires action code settings
+    guard let actionCodeSettings = configuration
+      .emailLinkSignInActionCodeSettings else {
+      throw AuthServiceError
+        .notConfiguredActionCodeSettings(
+          "ActionCodeSettings has not been configured for `AuthConfiguration.emailLinkSignInActionCodeSettings`"
+        )
+    }
+    return actionCodeSettings
+  }
+
+  public func updateAuthenticationState() {
+    authenticationState =
+      (currentUser == nil || currentUser?.isAnonymous == true)
+        ? .unauthenticated
+        : .authenticated
+  }
+
+  public func signOut() async throws {
+    try await auth.signOut()
+    updateAuthenticationState()
+  }
+
+  public func linkAccounts(credentials credentials: AuthCredential) async throws {
+    authenticationState = .authenticating
+    do {
+      try await currentUser?.link(with: credentials)
+      updateAuthenticationState()
+    } catch {
+      authenticationState = .unauthenticated
+      throw error
+    }
+  }
+
+  public func signIn(credentials credentials: AuthCredential) async throws {
+    authenticationState = .authenticating
+    if currentUser?.isAnonymous == true, configuration.shouldAutoUpgradeAnonymousUsers {
+      try await linkAccounts(credentials: credentials)
+    } else {
+      do {
+        try await auth.signIn(with: credentials)
+        signedInCredential = credentials
+        updateAuthenticationState()
+      } catch {
+        authenticationState = .unauthenticated
+        throw error
+      }
+    }
+  }
+
+  func sendEmailVerification() async throws {
+    if currentUser != nil {
+      do {
+        // TODO: - can use set user action code settings?
+        try await currentUser!.sendEmailVerification()
+      } catch {
+        throw error
+      }
+    }
+  }
+}
+
+// MARK: - User API
+
+extension Date {
+  func isWithinPast(minutes: Int) -> Bool {
+    let calendar = Calendar.current
+    guard let timeAgo = calendar.date(byAdding: .minute, value: -minutes, to: Date()) else {
+      return false
+    }
+    return self >= timeAgo && self <= Date()
+  }
+}
+
+public extension AuthService {
+  func reauthenticate() async throws {
+    if let user = auth.currentUser, let credential = signedInCredential {
+      try await user.reauthenticate(with: credential)
+    }
+  }
+
+  func deleteUser() async throws {
+    do {
+      if let user = auth.currentUser, let lastSignInDate = user.metadata.lastSignInDate {
+        let needsReauth = !lastSignInDate.isWithinPast(minutes: 5)
```

Comment:
Instead of making assumptions about the grace period, it'd be better to make the call and handle the error.

See https://github.com/peterfriese/IdentityKit/blob/a9383206ef3dce073e3b64b6828ccb8d3b77315b/Sources/IdentityKit/Service/AccountService.swift#L68